[SPARK-15616][SQL] Hive table supports partition pruning in JoinSelection

advancedxy opened the pull request:
### What changes were proposed in this pull request?
A new optimizer strategy called `PruneHiveTablePartitions` is added, which calculates table size as the total size of pruned partitions. Thus, Spark planner can pick up `BroadcastJoin` if the size of pruned partitions is under broadcast join threshold.

### Why are the changes needed?
This is a performance improvement.


### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Added unit tests.

This is based on #18193, credits should go to @lianhuiwang.

advancedxy committed:
[SPARK-15616][SQL] Hive table supports partition pruning in JoinSelection

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented:
ok to test

cloud-fan commented:
add to whitelist

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -232,6 +234,63 @@ case class RelationConversions(
+          predicate.references.subsetOf(partitionSet)
+      }
+      val conf = session.sessionState.conf
+      if (pruningPredicates.nonEmpty && conf.fallBackToHdfsForStatsEnabled &&
```
Why we need to check `conf.fallBackToHdfsForStatsEnabled`?

advancedxy replied:
We should only get size from HDFS if `conf.fallBackToHdfsForStatsEnabled`? Since it could be a time-consuming operation. 

Though, this condition should probably be pushed down to before the `CommandUtils.calculateLocationSize` call

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -232,6 +234,63 @@ case class RelationConversions(
+      val conf = session.sessionState.conf
+      if (pruningPredicates.nonEmpty && conf.fallBackToHdfsForStatsEnabled &&
+        conf.metastorePartitionPruning) {
+        val prunedPartitions = session.sharedState.externalCatalog.listPartitionsByFilter(
```
so we list partitions here just to get statistic, and we will list partitions again at runtime?

advancedxy replied:
Currently, yes. And I didn't find an easy way to avoid the list partitions operation.

cloud-fan replied:
Then this may make the performance worse. How about we keep the listed partitions in `HiveTableRelation`?

advancedxy replied:
> How about we keep the listed partitions in HiveTableRelation?

This is a good one. However, we may have to add two fields: `pruningFilters: Seq[Expression]` and `prunedPartitions: Seq[CatalogTablePartition]`, and I believe they are complicating the `HiveTableRelation`. Another things is that `HiveTableRelation` might be copied multiple times, we may lost the `prunedPartitions` field by accident if not copied carefully. 

If that's not the problem, storing listed partitions in `HiveTableRelation` is a good choice. 
WDYT @cloud-fan?

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16354/
Test FAILed.

SparkQA commented:
**[Test build #111297 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111297/testReport)** for PR 25919 at commit [`ffb7168`](https://github.com/apache/spark/commit/ffb71684bd7c2b48ab10f0acbec3df89b9b3f6b6).

SparkQA commented:
**[Test build #111297 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111297/testReport)** for PR 25919 at commit [`ffb7168`](https://github.com/apache/spark/commit/ffb71684bd7c2b48ab10f0acbec3df89b9b3f6b6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PruneHiveTablePartitions(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111297/
Test FAILed.

advancedxy committed:
Keep pruned partition in HiveTableRelation to avoid multiple calls to
hive meta store.

SparkQA commented:
**[Test build #112331 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112331/testReport)** for PR 25919 at commit [`c054d22`](https://github.com/apache/spark/commit/c054d22dda61883737c217459e36797f0bc796ba).

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17310/
Test PASSed.

advancedxy commented:
@cloud-fan Now pruned partitions are cached in HiveTableRelation, what do you think about current approach ?

SparkQA commented:
**[Test build #112332 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112332/testReport)** for PR 25919 at commit [`e744da5`](https://github.com/apache/spark/commit/e744da59aad15e8d1f5c7582fab195ddfde0a18c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17311/
Test PASSed.

SparkQA commented:
**[Test build #112331 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112331/testReport)** for PR 25919 at commit [`c054d22`](https://github.com/apache/spark/commit/c054d22dda61883737c217459e36797f0bc796ba).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112331/
Test FAILed.

SparkQA commented:
**[Test build #112332 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112332/testReport)** for PR 25919 at commit [`e744da5`](https://github.com/apache/spark/commit/e744da59aad15e8d1f5c7582fab195ddfde0a18c).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112332/
Test FAILed.

advancedxy committed:
Fix compile error

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17313/
Test PASSed.

SparkQA commented:
**[Test build #112334 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112334/testReport)** for PR 25919 at commit [`12e1dc5`](https://github.com/apache/spark/commit/12e1dc56db32ee4c939e8ad8260d484a30c7b3f8).

SparkQA commented:
**[Test build #112334 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112334/testReport)** for PR 25919 at commit [`12e1dc5`](https://github.com/apache/spark/commit/12e1dc56db32ee4c939e8ad8260d484a30c7b3f8).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112334/
Test FAILed.

advancedxy committed:
Fix compile error

SparkQA commented:
**[Test build #112337 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112337/testReport)** for PR 25919 at commit [`b334e99`](https://github.com/apache/spark/commit/b334e9926cc810f73e2c41934021ef7b5ffe7128).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17316/
Test PASSed.

SparkQA commented:
**[Test build #112337 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112337/testReport)** for PR 25919 at commit [`b334e99`](https://github.com/apache/spark/commit/b334e9926cc810f73e2c41934021ef7b5ffe7128).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112337/
Test FAILed.

advancedxy committed:
Exclude normalizedFilters and prunedPartitions in canonicalization.

advancedxy committed:
Remove scalar subquery in partition expression.

SparkQA commented:
**[Test build #112356 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112356/testReport)** for PR 25919 at commit [`8d615f7`](https://github.com/apache/spark/commit/8d615f73fccd47f3690af4d4806f120e5c058d9e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17335/
Test PASSed.

SparkQA commented:
**[Test build #112356 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112356/testReport)** for PR 25919 at commit [`8d615f7`](https://github.com/apache/spark/commit/8d615f73fccd47f3690af4d4806f120e5c058d9e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112356/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -610,7 +610,10 @@ case class HiveTableRelation(
-    tableStats: Option[Statistics] = None) extends LeafNode with MultiInstanceRelation {
+    tableStats: Option[Statistics] = None,
+    @transient normalizedFilters: Seq[Expression] = Nil,
+    @transient prunedPartitions: Seq[CatalogTablePartition] = Nil)
```
How can we distinguish 0 partitions after pruning, and not being partition pruned?

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -231,6 +232,68 @@ case class RelationConversions(
+case class PruneHiveTablePartitions(
+  session: SparkSession) extends Rule[LogicalPlan] with PredicateHelper {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    case filter @ Filter(condition, relation: HiveTableRelation) if relation.isPartitioned =>
```
can we follow `PruneFileSourcePartitions`? I think we should also support `Filter(Project(HiveScan))`

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -231,6 +232,68 @@ case class RelationConversions(
+        val withStats = relation.tableMeta.copy(
+          stats = Some(CatalogStatistics(sizeInBytes = BigInt(sizeInBytes))))
+        val prunedHiveTableRelation = relation.copy(tableMeta = withStats,
+          normalizedFilters = pruningPredicates, prunedPartitions = prunedPartitions)
```
Why do we need to keep `pruningPredicates`? IIUC the approach should be very simply:
1. this rule only changes `HiveTableRelation` to hold an optional partition list.
2. the `HiveTableScanExec` will get the partition list from `HiveTableRelation` or call `listPartitionsByFilter`.

advancedxy replied:
Due to [SPARK-24085](https://issues.apache.org/jira/browse/SPARK-24085), the pruningPredicates(we eliminate the subquery) could be different than the filters passed to `HiveTableScan`. So I keep the pruningPredicates, and only retrieves the `prunedPartitions` when `HiveTableScanExec`'s `pruningPartitionPredict` matches exactly with `HiveTableRelation`'s `normalizedFilters`.

The simplified solution occurred to me first, then I thought the filters could be different for some reason, and SPARK-24085 is an example, hence the proposed solution here.

advancedxy committed:
Address comments:
1. don't store pruningFilters in HiveTableRelation
2. follow PruneFilSourcePartitions's style to extract projections,
   predicates and hive relation
3. skip partition pruning if scalar subquery is involved.

SparkQA commented:
**[Test build #112679 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112679/testReport)** for PR 25919 at commit [`86a0d9c`](https://github.com/apache/spark/commit/86a0d9c489501e1e7b42351154973f21a838763b).

SparkQA commented:
**[Test build #112679 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112679/testReport)** for PR 25919 at commit [`86a0d9c`](https://github.com/apache/spark/commit/86a0d9c489501e1e7b42351154973f21a838763b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17636/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112679/
Test FAILed.

advancedxy committed:
Fix style issue.

SparkQA commented:
**[Test build #112707 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112707/testReport)** for PR 25919 at commit [`ecfbe4d`](https://github.com/apache/spark/commit/ecfbe4d975aeee1da38240753c99a6927a8aa690).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17655/
Test PASSed.

SparkQA commented:
**[Test build #112707 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112707/testReport)** for PR 25919 at commit [`ecfbe4d`](https://github.com/apache/spark/commit/ecfbe4d975aeee1da38240753c99a6927a8aa690).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112707/
Test FAILed.

advancedxy commented:
retest it please

advancedxy commented:
Gently ping @cloud-fan

maropu commented:
retest this please

maropu commented:
still `WIP`?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19169/
Test PASSed.

SparkQA commented:
**[Test build #114331 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114331/testReport)** for PR 25919 at commit [`ecfbe4d`](https://github.com/apache/spark/commit/ecfbe4d975aeee1da38240753c99a6927a8aa690).

SparkQA commented:
**[Test build #114331 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114331/testReport)** for PR 25919 at commit [`ecfbe4d`](https://github.com/apache/spark/commit/ecfbe4d975aeee1da38240753c99a6927a8aa690).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114331/
Test PASSed.

advancedxy commented:
> still `WIP`?

I think it's ready for review.

fuwhu commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala:
```diff
@@ -166,14 +166,14 @@ case class HiveTableScanExec(
         val normalizedFilters = partitionPruningPred.map(_.transform {
           case a: AttributeReference => originalAttributes(a)
         })
-        sparkSession.sessionState.catalog.listPartitionsByFilter(
```
@cloud-fan @maropu @advancedxy 
Since the rawPartitions are called by "prunePartitions(rawPartitions)" in doExecute method, it seems prunePartitions will filter out all irrelevant partitions using "boundPruningPred". Then why we still need to call listpartitionsByFilter here ?
Could you please help me understand this ? thanks a lot in advance.

fuwhu commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -231,6 +232,68 @@ case class RelationConversions(
+        !predicate.references.isEmpty && predicate.references.subsetOf(partitionSet)
+      }
+      // SPARK-24085: scalar subquery should be skipped for partition pruning
+      val hasScalarSubquery = pruningPredicates.exists(SubqueryExpression.hasSubquery)
```
I think it skip all subqueries instead of scalar subqueries.

fuwhu commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -231,6 +232,68 @@ case class RelationConversions(
+              rawDataSize.get
+            } else if (totalSize.isDefined && totalSize.get > 0L) {
+              totalSize.get
+            } else if (conf.fallBackToHdfsForStatsEnabled) {
```
Per the doc of the conf "spark.sql.statistics.fallBackToHdfs", it is only for non-partitioned hive table : 
"This flag is effective only for non-partitioned Hive tables."

advancedxy commented:
closed in favor of #26805
